packages/babel-generator/src/buffer.js

```diff
@@ -40,7 +39,7 @@ export default class Buffer {
 
     const map = this._map;
     const result = {
-      code: trimEnd(this._buf.join("")),
+      code: this._trimEnd(this._buf.join("")),
```

Comment:
To throw out a guess, the generator injects lots of newlines trying to make formatting match, so this probably keeps the end-of-file formatting the same even through the node type at the end of the file may differ.